HPCC-17935 Fix Meta Generator for Documentation

successar opened the pull request:
Signed-off-by: Sarthak Jain <successar@gmail.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

richardkchapman commented:
@ghalliday Please review

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: b485507beb2f170e32b00bbba2a87631280f580f
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 46.387 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |741 |741 |0 |
|test (thor) |636 |636 |0 |
|test (roxie) |769 |769 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented on system/jlib/jptree.cpp:
```diff
@@ -6028,74 +6028,33 @@ static const char * skipWhitespace(const char * text)
 
 void extractJavadoc(IPropertyTree * result, const char * text)
 {
-    //Skip a leading blank line
```
Why have you removed this function? Probably better would be to replace the call instead, since this is now doing nothing.

successar replied:
Processing of documentation is being done in python ecldoc code (since I have added more markup to it, which will make it difficult to process in c++ and in this way, we will have single area where docstring processing occurs). I will replace the call with code directly in hqldesc. Should I remove this function then from here ?

ghalliday replied:
You need to replace the call in hqlparse.cpp:

    extractJavadoc(tree, javaDocComment.str());

with

   tree->addProp("content", javaDocComment.str());

I would retain that function for the moment - and I also need to know if Gordon is currently using the extracted information.

ghalliday reviewed (changes requested):
@successar generally the changes look good.  There are several style comments to make sure that it fits in with the style of the rest of the project.

Long term a potential concern is the size of the data being generated.  At the moment one of my sample queries is generating a meta file 118Mb in size.

ghalliday commented:
I am also getting entries that look like this:

  <Import body="0" end="10" fullname="ut" inherit_type="local" line="2" name="ut" ref="ut" remotescope="1" start="10">
   <Type>attribute</Type>
  </Import>

I don't think the <Type> should be present.

ghalliday commented on ecl/hql/hqldesc.cpp:
```diff
@@ -90,9 +109,6 @@ static void expandRecordSymbolsMeta(IPropertyTree * metaTree, IHqlExpression * r
             {
                 IPropertyTree * field = metaTree->addPropTree("Field");
                 field->setProp("@name", str(cur->queryId()));
-                StringBuffer ecltype;
```
why remove the type of the field?

To correctly represent record types may require more though to prevent the file getting too large.

successar commented on ecl/hql/hqldesc.cpp:
```diff
@@ -113,15 +129,98 @@ static void expandRecordSymbolsMeta(IPropertyTree * metaTree, IHqlExpression * r
+    if (expr->hasAttribute(virtualAtom))
+        meta->setPropBool("@virtual", true);
+
+    if (expr->hasAttribute(interfaceAtom)){
```
Roger and I decided to put the development of type system once we have completed first iteration of ecldoc. Since Types can have complicated structures (for eg DATASET(rec_1), SET of STRING, etc), it will difficult to represent the relationships between different parts within the type correctly in an attribute.

ghalliday replied:
That is true.  You will also need to be careful about not duplicating information when it is shared by multiple definitions - .g, only expanding the fields in a record once.  Other complications are the introduction of a distributed meta file generation - where a single file is generated for each source file.  I'm not sure what other problems that will create.

successar commented on ecl/hql/hqldesc.cpp:
```diff
@@ -113,15 +129,98 @@ static void expandRecordSymbolsMeta(IPropertyTree * metaTree, IHqlExpression * r
+
+    if (param)
+    {
+        param->setProp("@name", str(cur->queryId()));
```
Again, I will add Types in params, fields, etc once we develop the type system representation.

successar commented on ecl/hql/hqldesc.cpp:
```diff
@@ -136,35 +235,42 @@ void expandSymbolMeta(IPropertyTree * metaTree, IHqlExpression * expr)
-            setNonZeroPropInt(def, "@start", symbol->getStartPos());
-            setNonZeroPropInt(def, "@body", symbol->getBodyPos());
-            setNonZeroPropInt(def, "@end", symbol->getEndPos());
+            def->setPropInt("@start", symbol->getStartPos());
```
We are using start and body pos to extract definition signatures out of ecl files. Sometimes, a definition start at beginning of file which has start pos 0. Therefore, including start = 0 provides consistency in later processing and helps in debugging. 

But if it is a problem, I can revert this to original form and add conditionals in my ecldoc python code.

ghalliday replied:
Possibly a good compromise would be to suppress them if the endPos is 0.  However it is also worth investigating why it is 0 for some items.

successar commented on ecl/hql/hqldesc.cpp:
```diff
@@ -136,35 +235,42 @@ void expandSymbolMeta(IPropertyTree * metaTree, IHqlExpression * expr)
+            def->setPropInt("@start", symbol->getStartPos());
+            def->setPropInt("@body", symbol->getBodyPos());
+            def->setPropInt("@end", symbol->getEndPos());
+            def->setProp("@source", str(expr->querySourcePath()));
```
we use sourcepath to read ecl file and extract signature for corresponding definition from it. While most definitions has source path of containing tag, some (specially inherited and overriden attributes) do have sources in other files. 
What I can do is only add source path only if it different from one in containing tag.

successar commented on ecl/hql/hqldesc.cpp:
```diff
@@ -136,35 +235,42 @@ void expandSymbolMeta(IPropertyTree * metaTree, IHqlExpression * expr)
+            def->setProp("Type", "type");
+        }
+        else {
+            def->setProp("Type", "attribute");
```
Again, we will add Type info later.

ghalliday commented:
Also, @GordonSmith do you use the structure of the javadoc that was previously generated?

@successar (to ensure it is logically assigned to you)

successar commented:
I am not sure what you mean by attribute entries. Can you give an example where they are being generated ?

ghalliday commented:
I don't understand the question about attribute entities - I can't see where I used that phrase.

It could be a comment that using getECLType() is better than "attribute".

successar commented:
The question was regarding following comment. I get it now.
"I am also getting entries that look like this:

  <Import body="0" end="10" fullname="ut" inherit_type="local" line="2" name="ut" ref="ut" remotescope="1" start="10">
   <Type>attribute</Type>
  </Import>

I don't think the <Type> should be present."

successar commented on ecl/hql/hqldesc.cpp:
```diff
@@ -190,6 +296,29 @@ void expandScopeSymbolsMeta(IPropertyTree * meta, IHqlScope * scope)
+    }
+
+
+    ForEachItemIn(i, symbols){
```
Are you talking about this function - https://github.com/hpcc-systems/HPCC-Platform/blob/master/ecl/hql/hqlexpr.hpp#L1012 . Actually, I tried to use it initially, but it takes a lookupcontext parameter. How would I go about getting it ?

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 42defbae0405b7da87ab5b20329a1f138c20eae3
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 45.442 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |741 |741 |0 |
|test (thor) |636 |636 |0 |
|test (roxie) |769 |769 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 2c882a2c74008ce9c1ccb60ef9fe6fd4743c6695
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 44.751 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |741 |741 |0 |
|test (thor) |636 |636 |0 |
|test (roxie) |769 |769 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented:
@successar That is looking good.  Remaining issues:

- Is Gordon going to have problems with @type changed to Type.  I doubt it, and will merge without his confirmation.
- Type of fields are removed.  Any reason not to keep it in (setting Type) rather than waiting for full change?
- Avoid duplicating sourcePath if it matches the parent.  It also needs more thought long term - really this should found be resolving the fullname, and not included here.
- restore the function in jptree.  It is possibly it may be deleted, but not yet.

Finally, some examples from the ecl/regress directory fail with an internal error:

minherit31, minherit31b, minherit32/32a/32b

successar committed:
HPCC-17935 Fix Meta Generator for Documentation

Signed-off-by: Sarthak Jain <successar@gmail.com>

successar commented:
@ghalliday 
- I have reverted back the code for types in field of record.
- I have removed the sourcepath from definitions and use fullname to resolve it.
- Reverted changes in jptree.
- Fixed the errors in minherit3* ecl files.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 9653b153e3731d28e03c42f3ce602a398f0ca092
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 42.469 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |741 |741 |0 |
|test (thor) |636 |636 |0 |
|test (roxie) |769 |769 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 5b5b464a7470a19501d29583404ce09ddc983af2
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 44.174 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |741 |741 |0 |
|test (thor) |636 |636 |0 |
|test (roxie) |769 |769 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 20d6b99acec484a406a5e6f5e7b76c21251a693c
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 42.405 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |741 |741 |0 |
|test (thor) |636 |636 |0 |
|test (roxie) |769 |769 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented:
@successar I don't see the change to implement "I have removed the sourcepath from definitions. I am using fullname to resolve it in my python code."

I'll grab your code and rerun my regression tests, but looks very close to being able to merge.

successar commented on ecl/hql/hqldesc.cpp:
```diff
@@ -136,35 +244,47 @@ void expandSymbolMeta(IPropertyTree * metaTree, IHqlExpression * expr)
+            def->setPropInt("@start", symbol->getStartPos());
+            def->setPropInt("@body", symbol->getBodyPos());
+            def->setPropInt("@end", symbol->getEndPos());
+            setFullNameProp(def, "@fullname", expr);
```
@ghalliday There used to be line here that set source path which i have removed now.

ghalliday commented:
Unfortunately the extra size of the meta data means I can no longer run my regression tests.  I will need to restructure the code to output the meta data earlier before I can finish testing this and merge it.

ghalliday commented:
I still have some concerns about the memory consumption  - but will merge as is.  Generating distributed meta information helps.
